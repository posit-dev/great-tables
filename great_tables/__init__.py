# Set version ----
from importlib_metadata import version as _v

__version__ = _v("great_tables")

del _v

# Main gt imports ----

from .gt import GT
from . import vals, loc, style
from ._styles import FromColumn as from_column
from ._helpers import (
    letters,
    LETTERS,
    px,
    pct,
    md,
    html,
    google_font,
    random_id,
    system_fonts,
    define_units,
    nanoplot_options,
)


__all__ = (
    "GT",
    "exibble",
    "letters",
    "LETTERS",
    "px",
    "pct",
    "md",
    "html",
    "google_font",
    "system_fonts",
    "define_units",
    "nanoplot_options",
    "random_id",
    "from_column",
    "vals",
    "loc",
    "style",
)


def __getattr__(k: str):
    # exibble dataset available on top-level module, but is a pandas DataFrame.
    # Since pandas is an optional dependency, we import exibble dynamically.
    if k == "exibble":
        from great_tables.data import exibble

        return exibble

    # allow the data submodule to be accessed directly, as if it were a top-level import
    if k == "data":
        import great_tables.data

        return great_tables.data

    raise AttributeError(f"cannot get attribute {k} from great_tables ({__file__})")
