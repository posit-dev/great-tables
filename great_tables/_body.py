from __future__ import annotations

from typing import TYPE_CHECKING

from ._tbl_data import copy_data

if TYPE_CHECKING:
    from ._gt_data import Body


def body_reassemble(body: Body) -> Body:
    # Note that this used to order the body based on groupings, but now that occurs in the
    # renderer itself.
    return body.__class__(copy_data(body.body))
