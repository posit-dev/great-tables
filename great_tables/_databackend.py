from __future__ import annotations

import importlib
from abc import ABCMeta
from types import ModuleType


def _load_class(mod_name: str, cls_name: str) -> type:
    # follow the path of mods if it is one
    listified_mod_path: list[str] = cls_name.split(".")

    sub_module_path: list[str] = [mod_name] + listified_mod_path[:-1]

    cur_mod: ModuleType = importlib.import_module(sub_module_path[0])
    _cls_name: str = listified_mod_path[-1]

    if len(sub_module_path) == 1:  # no submodules, just the module itself
        return getattr(cur_mod, _cls_name)

    # now, we must increment the module paths to import the final class
    remaining_mods: list[str] = sub_module_path[1:]
    for mod in remaining_mods:
        cur_mod: ModuleType = getattr(cur_mod, mod)

    return getattr(cur_mod, _cls_name)


class _AbstractBackendMeta(ABCMeta):
    def register_backend(cls, mod_name: str, cls_name: str):
        cls._backends.append((mod_name, cls_name))
        cls._abc_caches_clear()


class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented
