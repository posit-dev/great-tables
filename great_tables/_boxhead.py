from __future__ import annotations

from typing import TYPE_CHECKING

from ._locations import resolve_cols_c
from ._utils import _assert_list_is_subset
from ._tbl_data import SelectExpr
from ._text import BaseText

if TYPE_CHECKING:
    from ._types import GTSelf


def cols_label(
    self: GTSelf, cases: dict[str, str | BaseText] | None = None, **kwargs: str | BaseText
) -> GTSelf:
    """
    Relabel one or more columns.

    There are three important pieces to labelling:

    * Each argument has the form: {name in data} = {new label}.
    * Multiple columns may be given the same label.
    * Labels may use curly braces to apply special formatting, called unit notation.
      For example, "area ({{ft^2}})" would appear as "area (ftÂ²)".

    See [`define_units()`](`great_tables.define_units`) for details on unit notation.

    Parameters
    ----------
    cases
        A dictionary where the keys are column names and the values are the labels. Labels may use
        [`md()`](`great_tables.md`) or [`html()`](`great_tables.html`) helpers for formatting.

    **kwargs
        Keyword arguments to specify column labels. Each keyword corresponds to a column name, with
        its value indicating the new label.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Notes
    -----
    GT always selects columns using their name in the underlying data. This means that a column's
    label is purely for final presentation.

    Examples
    --------

    The example below relabels columns from the `countrypops` data to start with uppercase.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "Uganda"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Country Name",
            year="Year",
            population="Population"
        )
    )
    ```

    Note that we supplied the name of the column as the key, and the new label as the value.

    We can also use Markdown formatting for the column labels. In this example, we'll use
    `md("*Population*")` to make the label italicized.

    ```{python}
    from great_tables import GT, md
    from great_tables.data import countrypops

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Name",
            year="Year",
            population=md("*Population*")
        )
    )
    ```

    We can also use unit notation to format the column labels. In this example, we'll use
    `{{cm^3 molecules^-1 s^-1}}` for part of the label for the `OH_k298` column.

    ```{python}
    from great_tables import GT
    from great_tables.data import reactions
    import polars as pl

    reactions_mini = (
        pl.from_pandas(reactions)
        .filter(pl.col("cmpd_type") == "mercaptan")
        .select(["cmpd_name", "OH_k298"])
    )

    (
        GT(reactions_mini)
        .fmt_scientific("OH_k298")
        .sub_missing()
        .cols_label(
            cmpd_name="Compound Name",
            OH_k298="OH, {{cm^3 molecules^-1 s^-1}}",
        )
    )
    ```
    """
    from great_tables._helpers import UnitStr

    cases = cases if cases is not None else {}
    new_cases = cases | kwargs

    # If nothing is provided, return `data` unchanged
    if len(new_cases) == 0:
        return self

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()
    mod_columns = list(new_cases.keys())

    # Stop function if any of the column names specified are not in `cols_labels`
    # msg: "All column names provided must exist in the input `.data` table."
    _assert_list_is_subset(mod_columns, set_list=column_names)

    # Handle units syntax in labels (e.g., "Density ({{ppl / mi^2}})")
    new_kwargs: dict[str, UnitStr | str | BaseText] = {}

    for k, v in new_cases.items():
        if isinstance(v, str):
            unitstr_v = UnitStr.from_str(v)

            if len(unitstr_v.units_str) == 1 and isinstance(unitstr_v.units_str[0], str):
                new_kwargs[k] = unitstr_v.units_str[0]
            else:
                new_kwargs[k] = unitstr_v

        elif isinstance(v, BaseText):
            new_kwargs[k] = v

        else:
            raise ValueError(
                "Column labels must be strings or BaseText objects. Use `md()` or `html()` for formatting."
            )

    boxhead = self._boxhead._set_column_labels(new_kwargs)

    return self._replace(_boxhead=boxhead)


def cols_align(self: GTSelf, align: str = "left", columns: SelectExpr = None) -> GTSelf:
    """
    Set the alignment of one or more columns.

    The `cols_align()` method sets the alignment of one or more columns. The `align` argument
    can be set to one of `"left"`, `"center"`, or `"right"` and the `columns` argument can be
    used to specify which columns to apply the alignment to. If `columns` is not specified, the
    alignment is applied to all columns.

    Parameters
    ----------
    align
        The alignment to apply. Must be one of `"left"`, `"center"`, or `"right"`.
    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Examples
    --------
    Let's use the `countrypops` to create a small table. We can change the alignment of the
    `population` column with `cols_align()`. In this example, the column label and body cells of
    `population` will be aligned to the left.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "San Marino"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini, rowname_col="year", groupname_col="country_name")
        .cols_align(align="left", columns="population")
    )
    ```

    """

    # Throw if `align` is not one of the three allowed values
    if align not in ["left", "center", "right"]:
        raise ValueError("Align must be one of 'left', 'center', or 'right'.")

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()

    # Upgrade `columns` to a list if `columns` is a string and not None
    if isinstance(columns, str):
        columns = [columns]
        _assert_list_is_subset(columns, set_list=column_names)
    elif columns is None:
        columns = column_names

    sel_cols = resolve_cols_c(data=self, expr=columns)

    # Set the alignment for each column
    return self._replace(_boxhead=self._boxhead._set_column_aligns(sel_cols, align=align))
