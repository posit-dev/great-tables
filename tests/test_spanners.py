import pandas as pd
import polars as pl
import polars.selectors as cs
import pytest
from great_tables import GT, exibble
from great_tables._gt_data import Boxhead, ColInfo, ColInfoTypeEnum, SpannerInfo, Spanners
from great_tables._helpers import UnitStr
from great_tables._spanners import (
    _validate_sel_cols,
    cols_hide,
    cols_move,
    cols_move_to_end,
    cols_move_to_start,
    empty_spanner_matrix,
    spanners_print_matrix,
    tab_spanner,
)
from great_tables._utils_render_html import _get_table_defs


@pytest.fixture
def spanners() -> Spanners:
    return Spanners(
        [
            SpannerInfo(spanner_id="a", spanner_level=0, vars=["col1"], built="A"),
            SpannerInfo(spanner_id="b", spanner_level=1, vars=["col2"], built="B"),
        ]
    )


@pytest.fixture
def boxhead() -> Boxhead:
    return Boxhead(
        [
            ColInfo(var="col1"),
            ColInfo(var="col2"),
            ColInfo(var="col3"),
            ColInfo(var="col4", type=ColInfoTypeEnum.hidden),
        ]
    )


def test_spanners_next_level_above_first(spanners: Spanners):
    assert spanners.next_level(["col1"]) == 1


def test_spanners_next_level_above_second(spanners: Spanners):
    assert spanners.next_level(["col2"]) == 2


def test_spanners_next_level_unique(spanners: Spanners):
    assert spanners.next_level(["col3"]) == 0


def test_spanners_print_matrix(spanners: Spanners, boxhead: Boxhead):
    mat, vars = spanners_print_matrix(spanners, boxhead)
    assert vars == ["col1", "col2", "col3"]
    assert mat == [
        {"col1": None, "col2": "B", "col3": None},
        {"col1": "A", "col2": None, "col3": None},
        {"col1": "col1", "col2": "col2", "col3": "col3"},
    ]


def test_spanners_print_matrix_arg_omit_columns_row(spanners: Spanners, boxhead: Boxhead):
    mat, vars = spanners_print_matrix(spanners, boxhead, omit_columns_row=True)
    assert vars == ["col1", "col2", "col3"]
    assert mat == [
        {"col1": None, "col2": "B", "col3": None},
        {"col1": "A", "col2": None, "col3": None},
    ]


def test_spanners_print_matrix_arg_include_hidden(spanners: Spanners, boxhead: Boxhead):
    mat, vars = spanners_print_matrix(spanners, boxhead, include_hidden=True)
    assert vars == ["col1", "col2", "col3", "col4"]
    assert mat == [
        {"col1": None, "col2": "B", "col3": None, "col4": None},
        {"col1": "A", "col2": None, "col3": None, "col4": None},
        {"col1": "col1", "col2": "col2", "col3": "col3", "col4": "col4"},
    ]


def test_spanners_print_matrix_exclude_stub():
    """spanners_print_matrix omits a selected column if it's in the stub."""
    info = SpannerInfo(spanner_id="a", spanner_level=0, vars=["x", "y"], built="A")
    spanners = Spanners([info])
    boxh = Boxhead([ColInfo(var="x"), ColInfo(var="y", type=ColInfoTypeEnum.stub)])

    mat, vars = spanners_print_matrix(spanners, boxh, omit_columns_row=True)
    assert vars == ["x"]
    assert mat == [{"x": "A"}]


def test_empty_spanner_matrix():
    mat, vars = empty_spanner_matrix(["a", "b"], omit_columns_row=False)

    assert vars == ["a", "b"]
    assert mat == [{"a": "a", "b": "b"}]


def test_empty_spanner_matrix_arg_omit_columns_row():
    mat, vars = empty_spanner_matrix(["a", "b"], omit_columns_row=True)

    assert vars == ["a", "b"]
    assert mat == []


def test_tab_spanners_with_columns():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    src_gt = GT(df)

    dst_span = SpannerInfo("a_spanner", 0, "a_spanner", vars=["b", "a"])

    new_gt = tab_spanner(src_gt, "a_spanner", columns=["b", "a"], gather=False)
    assert len(new_gt._spanners)
    assert new_gt._spanners[0] == dst_span


def test_tab_spanners_with_spanner_ids():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    src_gt = GT(df)

    # we'll be testing for this second spanner added
    dst_span = SpannerInfo("b_spanner", 1, "b_spanner", vars=["c", "b", "a"])

    gt_with_span = tab_spanner(src_gt, "a_spanner", columns=["b", "a"], gather=False)

    new_gt = tab_spanner(gt_with_span, "b_spanner", spanners="a_spanner", columns=["c"])

    assert len(new_gt._spanners) == 2
    assert new_gt._spanners[1] == dst_span


def test_tab_spanner_units_text():
    df = pd.DataFrame({"x": [1.234, 2.345], "y": [3.456, 4.567], "z": [5.678, 6.789]})

    gt = GT(df).tab_spanner(label="Area ({{m^2}})", columns=["x", "y"])

    spanner = gt._spanners[0]

    assert isinstance(spanner.spanner_label, UnitStr)


def test_tab_spanners_overlap():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    src_gt = GT(df)

    # we'll be testing for this second spanner added
    dst_span = SpannerInfo("b_spanner", 0, "b_spanner", vars=["b"])

    new_gt = src_gt.tab_spanner("a_spanner", columns=["a"], gather=False).tab_spanner(
        "b_spanner", columns=["b"]
    )

    assert len(new_gt._spanners) == 2
    assert new_gt._spanners[1] == dst_span


def test_multiple_spanners_above_one():
    from great_tables import GT, exibble

    gt = (
        GT(exibble, rowname_col="row", groupname_col="group")
        .tab_spanner("A", ["num", "char", "fctr"])
        .tab_spanner("B", ["fctr"])
        .tab_spanner("C", ["num", "char"])
        .tab_spanner("D", ["fctr", "date", "time"])
        .tab_spanner("E", spanners=["B", "C"])
    )

    # Assert that the spanners have been added in the correct
    # format and in the correct levels

    assert len(gt._spanners) == 5
    assert gt._spanners[0] == SpannerInfo("A", 0, "A", vars=["num", "char", "fctr"])
    assert gt._spanners[1] == SpannerInfo("B", 1, "B", vars=["fctr"])
    assert gt._spanners[2] == SpannerInfo("C", 1, "C", vars=["num", "char"])
    assert gt._spanners[3] == SpannerInfo("D", 2, "D", vars=["fctr", "date", "time"])
    assert gt._spanners[4] == SpannerInfo("E", 3, "E", vars=["fctr", "num", "char"])


def test_tab_spanners_with_gather():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    src_gt = GT(df)

    new_gt = tab_spanner(src_gt, "a_spanner", columns=["a", "c"], gather=True)

    assert len(new_gt._spanners) == 1
    assert [col.var for col in new_gt._boxhead] == ["a", "c", "b"]


def test_cols_width_partial_set():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "10px"})

    assert gt_tbl._boxhead[0].column_width == "10px"
    assert gt_tbl._boxhead[1].column_width == None
    assert gt_tbl._boxhead[2].column_width == None


def test_cols_width_fully_set():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "10px", "b": "20px", "c": "30px"})

    assert gt_tbl._boxhead[0].column_width == "10px"
    assert gt_tbl._boxhead[1].column_width == "20px"
    assert gt_tbl._boxhead[2].column_width == "30px"


def test_cols_width_mix_cases_kwargs():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "10px"}, **{"b": "20px"}, c="30px")

    assert gt_tbl._boxhead[0].column_width == "10px"
    assert gt_tbl._boxhead[1].column_width == "20px"
    assert gt_tbl._boxhead[2].column_width == "30px"


def test_cols_width_partial_set_pct():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "20%"})

    assert gt_tbl._boxhead[0].column_width == "20%"
    assert gt_tbl._boxhead[1].column_width == None
    assert gt_tbl._boxhead[2].column_width == None


def test_cols_width_fully_set_pct():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "20%", "b": "20%", "c": "60%"})

    assert gt_tbl._boxhead[0].column_width == "20%"
    assert gt_tbl._boxhead[1].column_width == "20%"
    assert gt_tbl._boxhead[2].column_width == "60%"


def test_cols_width_fully_set_pct_2():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "10%", "b": "10%", "c": "40%"})

    assert gt_tbl._boxhead[0].column_width == "10%"
    assert gt_tbl._boxhead[1].column_width == "10%"
    assert gt_tbl._boxhead[2].column_width == "40%"


def test_cols_width_html_colgroup():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "10px", "b": "20px", "c": "30px"})

    tbl_built = gt_tbl._build_data(context="html")
    table_defs = _get_table_defs(tbl_built)

    assert (
        str(table_defs["table_colgroups"])
        == '<colgroup>\n  <col style="width:10px;"/>\n  <col style="width:20px;"/>\n  <col style="width:30px;"/>\n</colgroup>'
    )


def test_cols_width_html_colgroup_hidden():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df).cols_width({"a": "10px", "b": "20px", "c": "30px"}).cols_hide(columns="b")

    tbl_built = gt_tbl._build_data(context="html")
    table_defs = _get_table_defs(tbl_built)

    assert (
        str(table_defs["table_colgroups"])
        == '<colgroup>\n  <col style="width:10px;"/>\n  <col style="width:30px;"/>\n</colgroup>'
    )


def test_cols_width_html_colgroup_stub():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]})
    gt_tbl = GT(df, rowname_col="b").cols_width({"a": "10px", "b": "20px", "c": "30px"})

    tbl_built = gt_tbl._build_data(context="html")
    table_defs = _get_table_defs(tbl_built)

    assert (
        str(table_defs["table_colgroups"])
        == '<colgroup>\n  <col style="width:20px;"/>\n  <col style="width:10px;"/>\n  <col style="width:30px;"/>\n</colgroup>'
    )


def test_cols_width_html_colgroup_complex():
    df = pd.DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8], "e": [9, 10]})
    gt_tbl = (
        GT(df, rowname_col="c")
        .cols_move_to_start(columns="d")
        .cols_hide(columns="a")
        .cols_width({"a": "10px", "b": "20px", "c": "30px", "d": "40px", "e": "50px"})
    )

    tbl_built = gt_tbl._build_data(context="html")
    table_defs = _get_table_defs(tbl_built)

    assert (
        str(table_defs["table_colgroups"])
        == '<colgroup>\n  <col style="width:30px;"/>\n  <col style="width:40px;"/>\n  <col style="width:20px;"/>\n  <col style="width:50px;"/>\n</colgroup>'
    )


@pytest.mark.parametrize("DF, columns", [(pd.DataFrame, "a"), (pl.DataFrame, cs.starts_with("a"))])
def test_cols_move_single_col(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_move(src_gt, columns=columns, after="b")
    assert [col.var for col in new_gt._boxhead] == ["b", "a", "c", "d"]


@pytest.mark.parametrize(
    "DF, columns",
    [(pd.DataFrame, ["a", "d"]), (pl.DataFrame, cs.starts_with("a") | cs.ends_with("d"))],
)
def test_cols_move_multi_cols(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_move(src_gt, columns=columns, after="b")
    assert [col.var for col in new_gt._boxhead] == ["b", "a", "d", "c"]


@pytest.mark.parametrize("DF, columns", [(pd.DataFrame, "c"), (pl.DataFrame, cs.starts_with("c"))])
def test_cols_move_to_start_single_col(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_move_to_start(src_gt, columns=columns)
    assert [col.var for col in new_gt._boxhead] == ["c", "a", "b", "d"]


@pytest.mark.parametrize(
    "DF, columns",
    [(pd.DataFrame, ["c", "d"]), (pl.DataFrame, cs.starts_with("c") | cs.ends_with("d"))],
)
def test_cols_move_to_start_multi_cols(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_move_to_start(src_gt, columns=columns)
    assert [col.var for col in new_gt._boxhead] == ["c", "d", "a", "b"]


@pytest.mark.parametrize("DF, columns", [(pd.DataFrame, "c"), (pl.DataFrame, cs.starts_with("c"))])
def test_cols_move_to_end_single_col(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_move_to_end(src_gt, columns=columns)
    assert [col.var for col in new_gt._boxhead] == ["a", "b", "d", "c"]


@pytest.mark.parametrize(
    "DF, columns",
    [(pd.DataFrame, ["a", "c"]), (pl.DataFrame, cs.starts_with("a") | cs.ends_with("c"))],
)
def test_cols_move_to_end_multi_cols(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_move_to_end(src_gt, columns=columns)
    assert [col.var for col in new_gt._boxhead] == ["b", "d", "a", "c"]


@pytest.mark.parametrize("DF, columns", [(pd.DataFrame, "c"), (pl.DataFrame, cs.starts_with("c"))])
def test_cols_hide_single_col(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_hide(src_gt, columns=columns)
    assert [col.var for col in new_gt._boxhead if col.visible] == ["a", "b", "d"]


@pytest.mark.parametrize(
    "DF, columns",
    [(pd.DataFrame, ["a", "d"]), (pl.DataFrame, cs.starts_with("a") | cs.ends_with("d"))],
)
def test_cols_hide_multi_cols(DF, columns):
    df = DF({"a": [1, 2], "b": [3, 4], "c": [5, 6], "d": [7, 8]})
    src_gt = GT(df)
    new_gt = cols_hide(src_gt, columns=columns)
    assert [col.var for col in new_gt._boxhead if col.visible] == ["b", "c"]


def test_validate_sel_cols():
    sel_cols = ["a", "b", "c"]
    col_vars = ["a", "b", "c", "d"]
    _validate_sel_cols(sel_cols, col_vars)


def test_validate_sel_cols_raises():
    sel_cols = []
    col_vars = ["a", "b", "c", "d"]
    with pytest.raises(Exception) as exc_info:
        _validate_sel_cols(sel_cols, col_vars)

    assert "No columns selected." in exc_info.value.args[0]

    sel_cols = ["a", "b", "c", "x"]
    with pytest.raises(ValueError) as exc_info:
        _validate_sel_cols(sel_cols, col_vars)

    assert (
        "All `columns` must exist and be visible in the input `data` table."
        in exc_info.value.args[0]
    )
