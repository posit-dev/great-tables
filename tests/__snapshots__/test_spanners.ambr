# serializer version: 1
# name: test_spanners_styled
  '''
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">
      <span>&nbsp</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="1" scope="col">
      <span>&nbsp;</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="2" style="background-color: #1133FF;" scope="colgroup">
      <span class="gt_column_spanner">C</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="1" scope="col">
      <span>&nbsp;</span>
    </th>
  </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">
      <span>&nbsp</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="1" scope="col">
      <span>&nbsp;</span>
    </th>
    <th class="gt_center gt_columns_bottom_border gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="3" style="background-color: #1133FF;" scope="colgroup">
      <span class="gt_column_spanner">B</span>
    </th>
  </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="2" colspan="1" scope="col" id=""></th>
    <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="3" style="background-color: #33BB88;" scope="colgroup" id="A">
      <span class="gt_column_spanner">A</span>
    </th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="2" colspan="1" scope="col" id="d">d</th>
  </tr>
  <tr class="gt_col_headings">
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="a">a</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="b">b</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="c">c</th>
  </tr>
  '''
# ---
